### DISCUSSION TITLE
Computed fields slow down all requests

### DISCUSSION THREAD

**Discussion by sunscreem** [statamic/cms#7202] [Q&A] [answered]
Good chance I'm being dumb here as I'm new to Computed fields.

In a service provider I have the following:

```php
        Collection::computed('news_articles', 'reactions', function ($article) {

                return collect(config('site.reactions'))->reduce(
                    function ($output, $reaction, $key) use ($article) {
                        if ($article[$key . '_reactions']->count()) {
                            $output .= $reaction['emoji'] . $article[$key . '_reactions']->count();
                        }
                        return $output;
                    },
                    ''
                );
         
        });
 ```
        
This works great, but after a bunch of digging I've found out it's significantly slowing down my front end when I try and load the `news_articles` as, I guess, it's creating that field for every request. 

How can I only set that field for control panel requests? 

Also, any idea's why this would be generating 38,000+ events (I can see these are mostly Statamic\\Events\\UserBlueprintFound events in debugbar)

**Comment by sunscreem** [comment]
If the answer is that I should be caching that, how can I ensure that cache gets invalidated when things on the page change?

**Comment by jasonvarga** [comment, marked as answer]
Wrap it in a Blink or Cache so it's not executed multiple times. https://statamic.dev/computed-values#performance

```php
Collection::computed('news_articles', 'reactions', function ($article) {
  $key = 'article_reaction-'.$article->id();
  return Cache::rememberForever($key, function () { // or Blink::once()
    // ...
  });
});
```

You can invalidate it by listening for `EntrySaved`

```php
Event::listen(function (EntrySaved $event) {
  Cache::forget('article_reaction-'.$event->entry->id());
});
```

If you're using Blink, nothing is actually cached since it's only for that request. You don't need to invalidate it.

**Reply by sunscreem** [reply]
I suspected that was the answer. Thanks for confirming @jasonvarga. 

How can I be sure that cache (key) will be invalidated when the content changes? Manually invalidate it in my controller when setting the reactions? I think I just answered my own question.

**Reply by sunscreem** [reply]
Thanks again Jason. Nailed it.
